### PR TITLE
Fix priority of platform tags for manylinux

### PR DESCRIPTION
**Pull request by zanieb**
Closes https://github.com/astral-sh/uv/issues/2477

See also:
- #2489

### CODE REVIEW

File: crates/platform-tags/src/tags.rs [hunk 1]
```diff
@@ -538,34 +544,34 @@ mod tests {
...
-            "manylinux_2_25_x86_64",
-            "manylinux_2_26_x86_64",
-            "manylinux_2_27_x86_64",
-            "manylinux_2_28_x86_64",
-            "manylinux2010_x86_64",
-            "manylinux2014_x86_64",
             "manylinux1_x86_64",
+            "manylinux2014_x86_64",
+            "manylinux2010_x86_64",
+            "manylinux_2_28_x86_64",
```
Comment (bdice) [inline crates/platform-tags/src/tags.rs:550]: Just to make sure I understand this -- does this ordering mean that tags `manylinux1` would be preferred over `manylinux_2_28`?

Note that the "legacy" manylinux tags `manylinux1`, `manylinux2010`, and `manylinux2014` have corresponding glibc versions and should be prioritized like their corresponding "new glibc-style" tags.

https://peps.python.org/pep-0600/#legacy-manylinux-tags

> The existing manylinux tags are redefined as aliases for new-style tags:
>
>    manylinux1_x86_64 is now an alias for manylinux_2_5_x86_64
>    manylinux1_i686 is now an alias for manylinux_2_5_i686
>    manylinux2010_x86_64 is now an alias for manylinux_2_12_x86_64
>    manylinux2010_i686 is now an alias for manylinux_2_12_i686
>    manylinux2014_x86_64 is now an alias for manylinux_2_17_x86_64
>    manylinux2014_i686 is now an alias for manylinux_2_17_i686
>    manylinux2014_aarch64 is now an alias for manylinux_2_17_aarch64
>    manylinux2014_armv7l is now an alias for manylinux_2_17_armv7l
>    manylinux2014_ppc64 is now an alias for manylinux_2_17_ppc64
>    manylinux2014_ppc64le is now an alias for manylinux_2_17_ppc64le
>    manylinux2014_s390x is now an alias for manylinux_2_17_s390x


Concretely, we should expect this priority order:

`manylinux_2_28` > ... > `manylinux_2_17` == `manylinux2014` > `manylinux_2_16` > ... > `manylinux_2_12` == `manylinux2010` > `manylinux_2_11` > ... > `manylinux_2_5` == `manylinux1`

With the aliasing, I _think_ you want to prefer new-style over old style (which resolves any ambiguity in the `==` portions above). Typically wheels are tagged like `pyarrow-15.0.1-cp312-cp312-manylinux_2_17_x86_64.manylinux2014_x86_64.whl` and include both new-style and old-style tags if applicable.

Comment (bdice) [reply crates/platform-tags/src/tags.rs:550]: (Please correct me if I'm reading this wrong, I'm basing this on my recollection of `pip` behavior and a quick skim of https://peps.python.org/pep-0600/#package-installers)

File: crates/platform-tags/src/tags.rs [hunk 2]
```diff
@@ -485,6 +487,10 @@ fn compatible_tags(platform: &Platform) -> Result<Vec<String>, PlatformError> {
             )));
         }
     };
+
+    // Tags are constructed in increasing order priority, reverse on return
+    platform_tags.reverse();
```
Comment (charliermarsh) [inline crates/platform-tags/src/tags.rs:492]: I need to read up on it, but I don't think this is quite what we want. We need to change the ordering of the `manylinux2010_` and friends statements too.

File: crates/platform-tags/src/tags.rs (hunk 1 above)
Comment (zanieb) [reply crates/platform-tags/src/tags.rs:550]: Thanks for the details! I can futz with the ordering as you've described and I'll double check that we're matching the expected ordering.

**Comment by zanieb** [comment]
I fixed the ordering of the legacy abi tags, but I'm not sure where it's documented that `linux_x86_64` should take priority over `manylinux_<abi>_x86_64`. If someone has a link for that part of the specification I'd appreciate it — I can add it to the code.

**Comment by charliermarsh** [comment]
I think it might be the other way around, I think higher priority tags come first in that code, so `linux_x86_64` would be last? That's at least my read of the code, but I didn't test it.

**Comment by charliermarsh** [comment]
You can test by comparing to:

```python
from packaging import tags

for tag in tags.sys_tags(): print(tag)
```

File: crates/platform-tags/src/tags.rs [hunk 3]
```diff
@@ -358,8 +358,12 @@ fn compatible_tags(platform: &Platform) -> Result<Vec<String>, PlatformError> {
...
                 platform_tags.extend(
-                    (min_minor..=*minor).map(|minor| format!("manylinux_{major}_{minor}_{arch}")),
+                    (min_minor..=*minor)
+                        .map(|minor| format!("manylinux_{major}_{minor}_{arch}"))
+                        .rev(),
                 );
+                // Support legacy manylinux tags with lower priority
+                // <https://peps.python.org/pep-0600/#legacy-manylinux-tags>
                 if (min_minor..=*minor).contains(&12) {
                     platform_tags.push(format!("manylinux2010_{arch}"));
```
Comment (charliermarsh) [inline crates/platform-tags/src/tags.rs:368]: Doesn't `manylinux2014` need to be greater than `manylinux2010`, and also greater than `manylinux_2_12`? I think these legacy tags need to be interspersed.

Comment (zanieb) [reply crates/platform-tags/src/tags.rs:368]: Eek okay this was more wrong than I thought to start..

**Comment by zanieb** [comment]
Apologies for all the back and forth apparently this needed a more rigorous look :) 

I've addressed the last round of comments. I might look into a better way to test we match Python's behavior.

**Comment by zanieb** [comment]
It looks like our ordering of ABI tags is wrong, i.e. Python does

```

cp311-cp311-manylinux_2_28_x86_64
...
cp311-cp311-manylinux_2_17_x86_64
cp311-cp311-manylinux2014_x86_64
cp311-cp311-manylinux_2_16_x86_64
...
cp311-cp311-manylinux_2_12_x86_64
cp311-cp311-manylinux2010_x86_64
cp311-cp311-manylinux_2_11_x86_64
...
cp311-cp311-manylinux_2_5_x86_64
cp311-cp311-manylinux1_x86_64
cp311-cp311-linux_x86_64
cp311-abi3-manylinux_2_28_x86_64
...
cp311-abi3-manylinux_2_17_x86_64
cp311-abi3-manylinux2014_x86_64
cp311-abi3-manylinux_2_16_x86_64
...
cp311-abi3-manylinux_2_12_x86_64
cp311-abi3-manylinux2010_x86_64
cp311-abi3-manylinux_2_11_x86_64
...
cp311-abi3-manylinux_2_5_x86_64
cp311-abi3-manylinux1_x86_64
cp311-abi3-linux_x86_64
cp311-none-manylinux_2_28_x86_64
...
cp311-none-manylinux_2_17_x86_64
cp311-none-manylinux2014_x86_64
cp311-none-manylinux_2_16_x86_64
...
cp311-none-manylinux_2_12_x86_64
cp311-none-manylinux2010_x86_64
cp311-none-manylinux_2_11_x86_64
...
cp311-none-manylinux_2_5_x86_64
cp311-none-manylinux1_x86_64
cp311-none-linux_x86_64
cp310-abi3-manylinux_2_28_x86_64
...
cp310-abi3-manylinux_2_17_x86_64
cp310-abi3-manylinux2014_x86_64
cp310-abi3-manylinux_2_16_x86_64
cp310-abi3-manylinux_2_15_x86_64
...
cp310-abi3-manylinux_2_12_x86_64
cp310-abi3-manylinux2010_x86_64
cp310-abi3-manylinux_2_11_x86_64
...
cp310-abi3-manylinux_2_5_x86_64
cp310-abi3-manylinux1_x86_64
cp310-abi3-linux_x86_64
cp39-abi3-manylinux_2_28_x86_64
...
```

but we do

```
cp311_cp311_manylinux_2_28_x86_64
cp311_none_manylinux_2_28_x86_64
...
cp311_cp311_manylinux_2_17_x86_64
cp311_none_manylinux_2_17_x86_64
cp311_cp311_manylinux2014_x86_64
cp311_none_manylinux2014_x86_64
cp311_cp311_manylinux_2_16_x86_64
cp311_none_manylinux_2_16_x86_64
...
cp311_cp311_manylinux_2_12_x86_64
cp311_none_manylinux_2_12_x86_64
cp311_cp311_manylinux2010_x86_64
cp311_none_manylinux2010_x86_64
cp311_cp311_manylinux_2_11_x86_64
cp311_none_manylinux_2_11_x86_64
...
cp311_cp311_manylinux_2_5_x86_64
cp311_none_manylinux_2_5_x86_64
cp311_cp311_manylinux1_x86_64
cp311_none_manylinux1_x86_64
cp311_cp311_linux_x86_64
cp311_none_linux_x86_64
cp311_abi3_manylinux_2_28_x86_64
...
cp311_abi3_manylinux_2_17_x86_64
cp311_abi3_manylinux2014_x86_64
cp311_abi3_manylinux_2_16_x86_64
...
cp311_abi3_manylinux_2_12_x86_64
cp311_abi3_manylinux2010_x86_64
cp311_abi3_manylinux_2_11_x86_64
...
cp311_abi3_manylinux_2_5_x86_64
cp311_abi3_manylinux1_x86_64
cp311_abi3_linux_x86_64
cp310_abi3_manylinux_2_28_x86_64
...
```

**Comment by charliermarsh** [comment]
Oh interesting… they might be mutually exclusive such that it hasn’t mattered? Hard to reason about from my phone :D Thank you for digging into this!

**Comment by zanieb** [comment]
I'm no expert, just mindlessly matching Python's behavior at this point. See #2489 for the ABI changes.

I'll wait for @konstin to review both of these.

**Review by charliermarsh — APPROVED** [review]
Great!

**Review by charliermarsh — APPROVED** [review]
I’m comfortable signing off on this if you want to merge.

**Review by bdice — APPROVED** [review]
Thanks, this looks right to me too.

**Pull request merged**